### PR TITLE
Fix parsing of for-loops

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/76439

Introduced with https://github.com/dotnet/roslyn/pull/75632.  So only in 17.13p2.  So fixing this will prevent the regression from making it out.

The core issue here was an incorrect understanding i had with order-of-evaluation and storage of values for args when invoking methods.  Specifically, i thought that the following:

```
M(
   x = y(),
   Z(ref x));
```

Was evaluated effectively as:

```
xloc = y();
__z_val = Z(ref xloc);
M(xloc, __z_val);
```

But it is not.  rather, it is:

```
__x_val = x = y();
__z_val = M(ref x);
M(__x_val, __z_val);
```

In other words, i thought the call to M woudl see the value of 'x' *after* the call to Z, while it actually evaluates and stores that value prior to the call to Z.  

This mattered during parsing as we use that value to store skipped tokens onto.  So if we use the value prior to updating, we lose those skipped tokens, leading to an invalid incremental parse.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler @RikkiGibson  ptal

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -9181,23 +9181,27 @@ private ForStatementSyntax ParseForStatement(SyntaxList<AttributeListSyntax> att
             var (variableDeclaration, initializers) = eatVariableDeclarationOrInitializers();
 
             // Pulled out as we need to track this when parsing incrementors to place skipped tokens.
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9183]: It looks like this comment was specific to `secondSemicolonToken` before, is it still relevant now?

File: src/Compilers/CSharp/Test/Syntax/IncrementalParsing/IncrementalParsingTests.cs [hunk 2]
```diff
@@ -3666,6 +3666,33 @@ enum VirtualKey
...
+
+            var position1 = source.IndexOf("i =") + 1;
+            var position2 = source.IndexOf("i <") + 1;
+            var position3 = source.IndexOf("i++") + 1;
+
+            text = text.WithChanges(
+                new TextChange(new TextSpan(position1, 0), "n"),
+                new TextChange(new TextSpan(position2, 0), "n"),
+                new TextChange(new TextSpan(position3, 0), "n"));
+            tree = tree.WithChangedText(text);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/IncrementalParsing/IncrementalParsingTests.cs:3691]: It would make the test slightly easier to grasp if we asserted the contents of the resulting text here.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9183]: yes/  will move.

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9183]:
```suggestion
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -9181,23 +9181,27 @@ private ForStatementSyntax ParseForStatement(SyntaxList<AttributeListSyntax> att
...
-            SyntaxToken secondSemicolonToken;
+            var firstSemicolonToken = eatCommaOrSemicolon();
+            var condition = this.CurrentToken.Kind is not SyntaxKind.SemicolonToken and not SyntaxKind.CommaToken
+                ? this.ParseExpressionCore()
+                : null;
+            var secondSemicolonToken = eatCommaOrSemicolon();
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9188]:
```suggestion
            // Pulled out as we need to track this when parsing incrementors to place skipped tokens.
            var secondSemicolonToken = eatCommaOrSemicolon();
```

File: src/Compilers/CSharp/Test/Syntax/IncrementalParsing/IncrementalParsingTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/IncrementalParsing/IncrementalParsingTests.cs:3701]: done.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson this is ready for another pass.

**Review by RikkiGibson — APPROVED** [review]
